src/file-selector.ts

```diff
@@ -147,64 +147,63 @@ function fromDataTransferItem(
   return Promise.resolve(fwp);
 }
 
-// https://developer.mozilla.org/en-US/docs/Web/API/FileSystemEntry
-async function fromEntry(entry: any) {
-  return entry.isDirectory ? fromDirEntry(entry) : fromFileEntry(entry);
+async function fromEntry(
+  entry: FileSystemEntry,
+): Promise<FileWithPath | FileArray[]> {
+  if (isFileSystemDirectoryEntry(entry)) {
+    return fromDirectoryEntry(entry);
+  } else if (isFileSystemFileEntry(entry)) {
+    return fromFileEntry(entry);
+  }
+
+  return [];
```

Comment:
I prefer the shorthand conditional here (or anywhere else where it's possible).

Btw, `fromEntry` should never be called from anywhere else but the directory traversal func, so the entry must be either a file or a dir.

Reply:
If that's the case, then we should print a warning in the console or something, rather than failing silently (in the previous case it would have tried to read the entry as a file and likely would have failed?).

Still, I'm not a fan of function calls just for the sake of having the right type. The compiler should be smart enough to figure that out ... it's one of the reasons why I'm not a big fan of TypeScript anymore, but that's not in the scope of this discussion 😄 

Btw, what does the output look like (the js)? I'd imagine the compiler would optimize and remove the function call since it isn't necessary ...

Reply:
> If that's the case, then we should print a warning in the console or something, rather than failing silently (in the previous case it would have tried to read the entry as a file and likely would have failed?).

This is also unclear to me, MDN seems to list this explicitly, but the specification [says the following](https://wicg.github.io/entries-api/#entries) about entries:

> ### 2.3. Entries
> An entry is either a file entry or a directory entry.

Going through the commit history on MDN doesn't clear this up, as this content is part of the initial commit for this API and does not contain further details. But the specification [continues on](https://wicg.github.io/entries-api/#filesystementry) with code examples:

> **EXAMPLE 5**
> Handling an entry: 
> ```js
> function handleEntry(entry) {
>   console.log('name: ' + entry.name);
>   console.log('path: ' + entry.fullPath);
>   if (entry.isFile) {
>     console.log('... is a file');
>   } else if (entry.isDirectory) {
>     console.log('... is a directory');
>   }
> }
> ```

So this seems to imply we should still handle them separately, but again it is not being explicit. However there is never a note that a `FileSystemEntry` could be anything else but a directory or a file.

@inexorabletash as the editor of the specification could you shine a light on this? Is the MDN documentation here outright wrong?

All that said, I don't think we should log a warning in this case, as `file-selector` only knows how to handle files and directories, so filtering everything else out (whatever that might be) seems perfectly valid to me.

> Still, I'm not a fan of function calls just for the sake of having the right type. The compiler should be smart enough to figure that out. 

Totally agree with this, unfortunately the types used by TypeScript are generated from the Web IDL of the specification. In this case that definition does not create a union for the `FileSystemEntry` (either a file or directory), whether that is intentional is not clear. This makes it essentially impossible to determine the type without a predicate or a cast, and I'd still prefer a predicate.

Either way, I still think that a simple function call would be considered more than acceptable within this context. In terms of performance it is essentially negligible, as any action here is entirely IO bound, and it is a neatly isolated piece of logic.

> Btw, what does the output look like (the js)? I'd imagine the compiler would optimize and remove the function call since it isn't necessary ...

The TypeScript compiler doesn't do any such optimization, the only thing it will do is check if the types are actually valid and strip them out of the compiled result. You can see this in action in the [TypeScript Playground](https://www.typescriptlang.org/play/?target=99&jsx=0#code/IYZwngdgxgBAZgV2gFwJYHsLwE7oLYCiEy2YAFAFAwwCmxpAXDAGKoA2NAymCMjYfTAAaCgEomABVx5UIGgB5WHAOqpkACwnANMAD4t2NAILZswMAG0AugD4YAbyoxUcGGVkARVNhpRk6UiIScjpg0VEHJ2ofZARsLDhpLx8-ALAg0jJQ0lEAbicAX1o2OWdXdxAlGgyQwXDI6miaWPicfCqarLr86gKKJxi4rGt8voooTF5nEGTff0DBGABeNydssCYq7l5+GpFxWkXZAw5tvjxZ1IXg5bt1gDpPbzm0-PHJ5GmOxZWu4M3DGddnUmOtpicuDxzt8bksbGtBI9KoZckA). Any transformation that might occur by the TypeScript compiler is to support older browsers, which is disabled entirely in this project by always targeting `ESNext`, resulting only in type-stripped code.

Reply:
Thanks @jonkoops for taking the time to go through the spec and impl. logic. I think it's fine to guard for it the way you've done it.

The only thing is that if we fail silently and the user expects to see something, they'll be confused as to why the scenario where it's not a file/dir is ignored? Anyway, we'll deal with that if the issue comes up.